File: typescript/packages/core/src/utils/getTools.ts
import type { PluginBase } from "../classes/PluginBase";
import type { ToolBase } from "../classes/ToolBase";
import type { WalletClientBase } from "../classes/WalletClientBase";

export type GetToolsParams<TWalletClient extends WalletClientBase> = {
    wallet: TWalletClient;
    plugins?: (PluginBase<TWalletClient> | PluginBase<WalletClientBase>)[];
};
export async function getTools<TWalletClient extends WalletClientBase>({
    wallet,
    plugins = [],
}: GetToolsParams<TWalletClient>) {
    const tools: ToolBase[] = [];

    const chain = wallet.getChain();
    const coreTools = wallet.getCoreTools();

    for (const plugin of plugins) {
        if (!plugin.supportsChain(chain)) {
            console.warn(
                `Plugin ${plugin.name} does not support ${chain.type}${
                    "id" in chain ? ` chain id ${chain.id}` : ""
                }. Skipping.`,
            );
        }
        const pluginTools = await plugin.getTools(wallet);
        tools.push(...pluginTools);
    }

    return [...coreTools, ...tools];
}

File: python/src/goat-sdk/goat/utils/get_tools.py
from typing import List, Optional
from ..classes.plugin_base import PluginBase
from ..classes.tool_base import ToolBase
from ..classes.wallet_client_base import WalletClientBase


def get_tools(
    wallet: WalletClientBase, plugins: Optional[List[PluginBase]] = None
) -> List[ToolBase]:
    """Get all tools from the wallet and plugins."""
    tools: List[ToolBase] = []
    plugins = plugins or []

    chain = wallet.get_chain()
    core_tools = wallet.get_core_tools()

    for plugin in plugins:
        if not plugin.supports_chain(chain):
            chain_id = f" chain id {chain['id']}" if "id" in chain else ""
            print(
                f"Warning: Plugin {plugin.name} does not support {chain['type']}{chain_id}. Skipping."
            )
            continue

        plugin_tools = plugin.get_tools(wallet)
        tools.extend(plugin_tools)

    return [*core_tools, *tools]
